File: static/CWE-327/python/patched.py
# Remediation for CWE-327: Use of a Broken or Risky Cryptographic Algorithm
# Fix: Hash passwords with a per-user random salt using SHA-256 (PBKDF2) instead of MD5.

import hashlib
import os
from flask import Flask, request

app = Flask(__name__)


@app.route("/register", methods=["POST"])
def register():
    password = request.form.get("password")
    salt = os.urandom(16)
    hashed = hashlib.pbkdf2_hmac("sha256", password.encode(), salt, 200_000)
    return "stored hash: " + salt.hex() + ":" + hashed.hex()

File: static/CWE-327/javascript/patched.js
// Remediation for CWE-327: Use of a Broken or Risky Cryptographic Algorithm
// Fix: Hash passwords with bcrypt (a strong, salted, adaptive algorithm) instead of MD5.

const express = require('express');
const bcrypt = require('bcrypt');

const app = express();
app.use(express.json());

const SALT_ROUNDS = 12;

function hashPassword(password) {
  return bcrypt.hash(password, SALT_ROUNDS);
}

app.post('/register', async (req, res) => {
  const password = req.body.password;
  const hashed = await hashPassword(password);
  res.json({ passwordHash: hashed });
});

app.listen(3000);
